### PR TITLE
Add mouse button events to InputHandler

### PR DESCRIPTION
**Pull request by derekdreery**
Added mouse buttons to the InputHandler (exactly like keys) and run through rustfmt.

I have implemented something locally for mouse movements, but I'm not sure whether there's a better API than the one I'm using so I'm submitting it as a separate PR.

Blocks #186

### CODE REVIEW

**Comment by derekdreery** [comment]
OK cool :).

File: src/ecs/resources/input.rs [hunk 1]
```diff
@@ -34,6 +34,27 @@ impl<'a> Iterator for PressedMouseButtonsIterator<'a> {
     }
 }
 
+/// This struct holds state information about input devices.
+///
+/// For example, if a key is pressed on the keyboard, this struct will record
+/// that the key is pressed until it is released again. Usage requires pumping input events through
+/// this struct, using the following code in the [handle_events] method of [State].
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: I didn't even think about using reference links to methods in the documentation. Nice! 😄

File: src/ecs/resources/input.rs [hunk 2]
```diff
@@ -22,14 +22,52 @@ impl<'a> Iterator for PressedKeysIterator<'a> {
     }
 }
 
+/// Iterator that iterates through all currently pressed down mouse buttons
+pub struct PressedMouseButtonsIterator<'a> {
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Hmm, this name seems quite long and cumbersome. Perhaps we could trim this down to `PressedMouseButtons`, following the pattern used in the standard library (e.g. `std::collections::hash_map::Keys` as opposed to `std::collections::hash_map::KeysIterator`)?

File: src/ecs/resources/input.rs [hunk 3]
```diff
@@ -45,14 +83,30 @@ impl InputHandler {
...
                 Event::KeyboardInput(ElementState::Released, _, Some(key_code)) => {
                     self.pressed_keys.remove(&key_code);
                 }
-                Event::Focused(false) => self.pressed_keys.clear(),
+                Event::MouseInput(ElementState::Pressed, button) => {
+                    match self.pressed_mouse_buttons.entry(button) {
+                        Entry::Occupied(_) => {
+                            // as above
+                        }
+                        Entry::Vacant(entry) => {
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: If we are really only using one branch of the `match` statement, couldn't this entire statement be shortened down to:

```rust
if let Entry::Vacant(entry) = self.pressed_mouse_buttons.entry(button) {
    entry.insert(KeyQueryState::NotQueried);
}
```

File: src/ecs/resources/input.rs [hunk 4]
```diff
@@ -93,4 +148,41 @@ impl InputHandler {
     pub fn keys_once(&mut self, keys: &[VirtualKeyCode]) -> bool {
         keys.iter().any(|key| self.key_once(*key)) && self.keys_down(keys)
     }
+
+    /// Returns an iterator for all the pressed down mouse buttons
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Tiny nitpick, missing a period.

File: src/ecs/resources/input.rs [hunk 5]
```diff
@@ -93,4 +148,41 @@ impl InputHandler {
...
+    pub fn pressed_mouse_buttons(&self) -> PressedMouseButtonsIterator {
+        PressedMouseButtonsIterator { iterator: self.pressed_mouse_buttons.keys() }
+    }
+
+    /// Check if `button` is currently pressed
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Should be `/// Checks` for consistency with the general documentation style, also missing a period.

File: src/ecs/resources/input.rs [hunk 6]
```diff
@@ -93,4 +148,41 @@ impl InputHandler {
...
+    pub fn mouse_button_down(&self, button: MouseButton) -> bool {
+        self.pressed_mouse_buttons.contains_key(&button)
+    }
+
+    /// Check if all `buttons` are currently pressed
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Should be `/// Checks` for consistency with the general documentation style, also missing a period.

File: src/ecs/resources/input.rs [hunk 7]
```diff
@@ -93,4 +148,41 @@ impl InputHandler {
...
+            if *value == KeyQueryState::NotQueried {
+                *value = KeyQueryState::Queried;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// Checks if `buttons` are all currently pressed and haven't been called with
+    /// `mouse_button_once` or `mouse_buttons_once`
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Also missing a period.

File: src/ecs/resources/input.rs [hunk 8]
```diff
@@ -93,4 +148,41 @@ impl InputHandler {
...
+    pub fn mouse_button_down(&self, button: MouseButton) -> bool {
+        self.pressed_mouse_buttons.contains_key(&button)
+    }
+
+    /// Check if all `buttons` are currently pressed
+    pub fn mouse_buttons_down(&self, buttons: &[MouseButton]) -> bool {
+        buttons.iter().all(|btn| self.mouse_button_down(*btn))
+    }
+
+    /// Check if `button` is currently pressed and `mouse_button_once` or `mouse_buttons_once`
```
Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Should be `/// Checks` for consistency with the general documentation style.

File: src/ecs/resources/input.rs (hunk 2 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: Should I change `PressedKeys` to match?

Comment (ebkalderon) [inline src/ecs/resources/input.rs]: Yes please, that would be a good idea. Sorry for asking for you to do it, I should have caught it before merging. :persevere:

Comment (derekdreery) [inline src/ecs/resources/input.rs]: no problem! :)

Comment (derekdreery) [inline src/ecs/resources/input.rs]: It's a non-breaking change as it's not possible to construct a PressedKeysIterator in user code

**Comment by derekdreery** [comment]
Could you check my merge to make sure it keeps the comments how you would like them?

File: src/ecs/resources/input.rs (hunk 2 above)
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: Not true, that is a breaking change, however we are following a specific type of semver for sub 1.0.0 projects, this means that changes like these don't need a major bump.

File: src/ecs/resources/input.rs (hunk 3 above)
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: Yes, I don't know why I kept the `Occupied` arm in the first pr.

**Comment by derekdreery** [comment]
@ebkalderon I've made all the changes but I don't know if there's something I need to do to mark it ready for review again.

File: src/ecs/resources/input.rs (hunk 2 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: Oh yeah you're right because you might use it in a type signature.

File: src/ecs/resources/input.rs (hunk 3 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: Would you like me to remove that one as well?

**Comment by Emilgardis** [comment]
@derekdreery Please try to use this https://github.com/amethyst/amethyst/blob/develop/CONTRIBUTING.md#dealing-with-upstream-changes

In summary, do this to correctly merge changes for PRs

```bash
git reset HEAD~1
git rebase upstream/develop
# Now some files will probably be in a non-mergable state, do
# change these files (check with `git status`) and then do `git add <file>`
# then do `git rebase --continue``
```

**Comment by derekdreery** [comment]
OK I'll rebase

**Comment by derekdreery** [comment]
Rebased

**Comment by vitvakatu** [comment]
There are two possibilities: we broke something in `timing.rs`, or AppVeyor just caught rare bug.

**Comment by derekdreery** [comment]
ooh interesting - I'll have a look when I get chance.

**Comment by torkleyy** [comment]
@ebkalderon Should work now (reason was https://github.com/amethyst/amethyst/pull/204)

@derekdreery Could you please resolve the conflicting file?

**Comment by derekdreery** [comment]
rebased

File: src/ecs/resources/input.rs [hunk 9]
```diff
@@ -15,27 +15,65 @@ enum KeyQueryState {
...
+    pressed_mouse_buttons: HashMap<MouseButton, KeyQueryState>,
 }
 
 impl InputHandler {
     /// Creates a new input handler.
     pub fn new() -> InputHandler {
-        InputHandler { pressed_keys: HashMap::default() }
+        InputHandler {
+            pressed_keys: HashMap::new(),
+            pressed_mouse_buttons: HashMap::new(),
```
Comment (torkleyy) [inline src/ecs/resources/input.rs]: Think it should be `HashMap::default`.

File: src/ecs/resources/input.rs [hunk 10]
```diff
@@ -106,4 +155,45 @@ impl InputHandler {
...
+            // Should be safe
+            if *value == KeyQueryState::NotQueried {
+                *value = KeyQueryState::Queried;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// Checks if the all the given mouse buttons are being pressed and held down.
```
Comment (torkleyy) [inline src/ecs/resources/input.rs]: Not a native English speaker, but I think something like

> Checks if all the given mouse buttons were pressed and not yet queried

would be better

File: src/ecs/resources/input.rs [hunk 11]
```diff
@@ -106,4 +155,45 @@ impl InputHandler {
...
+            return false;
+        }
+        if let Some(value) = self.pressed_mouse_buttons.get_mut(&button) {
+            // Should be safe
+            if *value == KeyQueryState::NotQueried {
+                *value = KeyQueryState::Queried;
+                return true;
+            }
+        }
+        return false;
```
Comment (vitvakatu) [inline src/ecs/resources/input.rs]: Please change it to `false` only, without `return`, according to `clippy` policy. I'll change it anyway in the next review with `clippy`.

File: src/ecs/resources/input.rs (hunk 10 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: In the situation where the mouse button(s) were pressed, then released, and not queried, then the result of the above comment would be true (i.e. they were pressend and not yet queried), but the function would return false (as they have been released).

File: src/ecs/resources/input.rs (hunk 11 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: agree and changed

File: src/ecs/resources/input.rs (hunk 9 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: agree and changed

File: src/ecs/resources/input.rs (hunk 10 above)
Comment (torkleyy) [inline src/ecs/resources/input.rs]: ah right, what about

> are being pressed and were not queriet yet

?

File: src/ecs/resources/input.rs [hunk 12]
```diff
@@ -106,4 +155,45 @@ impl InputHandler {
...
+                *value = KeyQueryState::Queried;
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Checks if the all the given mouse buttons are being pressed and held down.
+    ///
+    /// If the `buttons` haven't been let go since the last `mouse_button_once()` query,
```
Comment (vitvakatu) [inline src/ecs/resources/input.rs]: Maybe "haven't been released"?

Comment (derekdreery) [inline src/ecs/resources/input.rs]: yeah that's better, I'll update.

**Comment by derekdreery** [comment]
np. 👍

File: src/ecs/resources/input.rs [hunk 13]
```diff
@@ -51,22 +89,33 @@ impl InputHandler {
                             // second `Pressed` event.
                         }
                         Entry::Vacant(entry) => {
-                            entry.insert(KeyQueryState::Queried);
+                            entry.insert(KeyQueryState::NotQueried);
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: I think this change should be cherry picked out as another commit. We should try to be clear on what our commits are doing, adding mouse button events != bug fixes

File: src/ecs/resources/input.rs [hunk 14]
```diff
@@ -51,22 +89,33 @@ impl InputHandler {
...
                         }
                     }
                 }
                 Event::KeyboardInput(ElementState::Released, _, Some(key_code)) => {
                     self.pressed_keys.remove(&key_code);
                 }
-                Event::Focused(false) => self.pressed_keys.clear(),
+                Event::MouseInput(ElementState::Pressed, button) => {
+                    if let Entry::Vacant(entry) = self.pressed_mouse_buttons.entry(button) {
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: Why is this an `if let` but L84-94 is not. I would consider changing the match into an `if let`. When I did this I didn't think much of `if let`s.

File: src/ecs/resources/input.rs [hunk 15]
```diff
@@ -101,9 +150,50 @@ impl InputHandler {
...
+    pub fn pressed_mouse_buttons(&self) -> PressedMouseButtons {
+        PressedMouseButtons { iterator: self.pressed_mouse_buttons.keys() }
+    }
+
+    /// Checks if `button` is currently pressed.
+    pub fn mouse_button_down(&self, button: MouseButton) -> bool {
+        self.pressed_mouse_buttons.contains_key(&button)
+    }
+
+    /// Checks if all `buttons` are currently pressed.
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: Cryptic, see L126

> /// Checks if all the given keys are being pressed.

i.e this should be

> /// Checks if all the given mouse buttons are being pressed.

File: src/ecs/resources/input.rs [hunk 16]
```diff
@@ -101,9 +150,50 @@ impl InputHandler {
...
     pub fn keys_once(&mut self, keys: &[VirtualKeyCode]) -> bool {
         keys.iter().any(|key| self.key_once(*key)) && self.keys_down(keys)
     }
+
+    /// Returns an iterator for all the pressed down mouse buttons.
+    pub fn pressed_mouse_buttons(&self) -> PressedMouseButtons {
+        PressedMouseButtons { iterator: self.pressed_mouse_buttons.keys() }
+    }
+
+    /// Checks if `button` is currently pressed.
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: Same as below,

> /// Checks if the given mouse button is being pressed.

File: src/ecs/resources/input.rs [hunk 17]
```diff
@@ -101,9 +150,50 @@ impl InputHandler {
...
+    /// Checks if the given mouse button is being pressed and held down.
+    ///
+    /// If `button` hasn't been released since the last `mouse_button_once()` query, this
+    /// function will return false.
+    pub fn mouse_button_once(&mut self, button: MouseButton) -> bool {
+        if !self.pressed_mouse_buttons.contains_key(&button) {
+            return false;
+        }
+        if let Some(value) = self.pressed_mouse_buttons.get_mut(&button) {
+            // Should be safe
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: There is no unwrap and we don't use any `unsafe {}` blocks here, comment should be removed.

File: src/ecs/resources/input.rs [hunk 18]
```diff
@@ -15,27 +15,65 @@ enum KeyQueryState {
...
-impl<'a> Iterator for PressedKeysIterator<'a> {
+impl<'a> Iterator for PressedKeys<'a> {
     type Item = &'a VirtualKeyCode;
     fn next(&mut self) -> Option<Self::Item> {
         self.iterator.next()
     }
 }
 
-/// Processes user input events.
+/// An iterator over all currently pressed down mouse buttons.
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: L17 should probably also be changed

Comment (derekdreery) [inline src/ecs/resources/input.rs]: 👍

File: src/ecs/resources/input.rs [hunk 19]
```diff
@@ -15,27 +15,65 @@ enum KeyQueryState {
...
+///     // ...
+///     let mut input_handler = world.write_resource::<InputHandler>();
+///     input_handler.update(events);
+///     // ...
+/// }
+/// ```
 #[derive(Default)]
 pub struct InputHandler {
     pressed_keys: HashMap<VirtualKeyCode, KeyQueryState>,
+    // use KeyQueryState for mouse button query state as well
```
Comment (derekdreery) [inline src/ecs/resources/input.rs]: I just wanted to emphaize that even though the name `KeyQueryState` is specific to keys, its structure is appropriate for mouse buttons as well. However, I'm happy to remove.

File: src/ecs/resources/input.rs (hunk 14 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: np

**Comment by derekdreery** [comment]
I've made all these changes. I'll pull out the cherry-picked change into another PR and marked this one as blocked. When that one is committed, it should be a trivial rebase on this.

**Comment by derekdreery** [comment]
Instead of marking this pr as blocked, I just removed the change.

**Comment by derekdreery** [comment]
Ok rebased on that bugfix.
